In [1]:
import numpy as np
import pandas as pd
import matplotlib as plt

# Heart Disease Prediction — Exploratory Data Analysis


## 1. Load the Dataset

In [4]:
data = pd.read_csv('../data/processed.cleveland.data', sep=',')

In [7]:
print(data.shape)

(302, 14)


In [8]:
data.head()

   63.0  1.0  1.0.1  145.0  233.0  1.0.2  2.0  150.0  0.0  2.3  3.0 0.0.1  \
0  67.0  1.0    4.0  160.0  286.0    0.0  2.0  108.0  1.0  1.5  2.0   3.0   
1  67.0  1.0    4.0  120.0  229.0    0.0  2.0  129.0  1.0  2.6  2.0   2.0   
2  37.0  1.0    3.0  130.0  250.0    0.0  0.0  187.0  0.0  3.5  3.0   0.0   
3  41.0  0.0    2.0  130.0  204.0    0.0  2.0  172.0  0.0  1.4  1.0   0.0   
4  56.0  1.0    2.0  120.0  236.0    0.0  0.0  178.0  0.0  0.8  1.0   0.0   

   6.0  0  
0  3.0  2  
1  7.0  1  
2  3.0  0  
3  3.0  0  
4  3.0  0  


In [11]:
data.columns = ["age", "sex", "cp", "trestbps","chol","fbs","restecg","thalach",
                "exang","oldpeak","slope","ca","thal","num",]
print(data.head(2))

    age  sex   cp  trestbps   chol  fbs  restecg  thalach  exang  oldpeak  \
0  67.0  1.0  4.0     160.0  286.0  0.0      2.0    108.0    1.0      1.5   
1  67.0  1.0  4.0     120.0  229.0  0.0      2.0    129.0    1.0      2.6   

   slope   ca thal  num  
0    2.0  3.0  3.0    2  
1    2.0  2.0  7.0    1  


In [12]:
data.info()

<class 'pandas.DataFrame'>
RangeIndex: 302 entries, 0 to 301
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       302 non-null    float64
 1   sex       302 non-null    float64
 2   cp        302 non-null    float64
 3   trestbps  302 non-null    float64
 4   chol      302 non-null    float64
 5   fbs       302 non-null    float64
 6   restecg   302 non-null    float64
 7   thalach   302 non-null    float64
 8   exang     302 non-null    float64
 9   oldpeak   302 non-null    float64
 10  slope     302 non-null    float64
 11  ca        302 non-null    str    
 12  thal      302 non-null    str    
 13  num       302 non-null    int64  
dtypes: float64(11), int64(1), str(2)
memory usage: 33.2 KB


In [13]:
data.dtypes

age         float64
sex         float64
cp          float64
trestbps    float64
chol        float64
fbs         float64
restecg     float64
thalach     float64
exang       float64
oldpeak     float64
slope       float64
ca              str
thal            str
num           int64
dtype: object

In [16]:
data.isna().sum()

age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          0
thal        0
num         0
dtype: int64

In [17]:
(data == "?").sum()

age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          4
thal        2
num         0
dtype: int64

## Missing Values and Data Types

### Observations

- `isna()` cannot recognize `?` as a missing value.
- We used `(data == "?").sum()` to identify the `?` values in the dataset.
- These values need to be converted into proper missing values before preprocessing.

## Feature Description

| Feature | Type | Description |
|---------|------|-------------|
| age | Numerical | Age in years |
| sex | Binary | Sex (1 = male, 0 = female) |
| cp | Categorical | Chest pain type (1 = typical angina, 2 = atypical angina, 3 = non-anginal pain, 4 = asymptomatic) |
| trestbps | Numerical | Resting blood pressure in mm Hg on admission to the hospital |
| chol | Numerical | Serum cholesterol level in mg/dl |
| fbs | Binary | Fasting blood sugar > 120 mg/dl (1 = true, 0 = false) |
| restecg | Categorical | Resting electrocardiographic results (0 = normal, 1 = ST-T wave abnormality, 2 = probable or definite left ventricular hypertrophy) |
| thalach | Numerical | Maximum heart rate achieved |
| exang | Binary | Exercise-induced angina (1 = yes, 0 = no) |
| oldpeak | Numerical | ST depression induced by exercise relative to rest |
| slope | Categorical | Slope of the peak exercise ST segment (1 = upsloping, 2 = flat, 3 = downsloping) |
| ca | Numerical | Number of major vessels (0–3) colored by fluoroscopy |
| thal | Categorical | Thalassemia result (3 = normal, 6 = fixed defect, 7 = reversible defect) |
| num | Categorical | Diagnosis of heart disease (0 = no presence of heart disease, 1–4 = presence of heart disease) |